src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,594 @@
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use
+     * @param provider
+     *     the provider to use for this key derivation
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if the specified provider does not support the specified KDF
+     *     algorithm
+     * @throws NoSuchProviderException
+     *     if the specified provider is not registered in the security provider
+     *     list
+     * @throws NullPointerException
+     *     if the algorithm or provider is {@code null}
+     */
+    public static KDF getInstance(String algorithm, String provider)
+        throws NoSuchAlgorithmException, NoSuchProviderException {
+        Objects.requireNonNull(provider, "provider may not be null");
+        try {
+            return getInstance(algorithm, null, provider);
+        } catch (InvalidAlgorithmParameterException e) {
+            throw new NoSuchAlgorithmException(
+                "Received an InvalidAlgorithmParameterException. Does this "
+                + "algorithm require KDFParameters?", e);
+        }
+    }
+
+    /**
+     * Returns a {@code KDF} object that implements the specified algorithm from
+     * the specified security provider.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use
+     * @param provider
+     *     the provider to use for this key derivation
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if the specified provider does not support the specified KDF
+     *     algorithm
+     * @throws NullPointerException
+     *     if the algorithm or provider is {@code null}
+     */
+    public static KDF getInstance(String algorithm, Provider provider)
+        throws NoSuchAlgorithmException {
+        Objects.requireNonNull(provider, "provider may not be null");
+        try {
+            return getInstance(algorithm, null, provider);
+        } catch (InvalidAlgorithmParameterException e) {
+            throw new NoSuchAlgorithmException(
+                "Received an InvalidAlgorithmParameterException. Does this "
+                + "algorithm require KDFParameters?", e);
+        }
+    }
+
+    /**
+     * Returns a {@code KDF} object that implements the specified algorithm and
+     * is initialized with the specified parameters.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use
+     * @param kdfParameters
+     *     the {@code KDFParameters} used to configure this KDF's algorithm or
+     *     {@code null} if no additional parameters are provided
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if no {@code Provider} supports a {@code KDFSpi} implementation for
+     *     the specified algorithm
+     * @throws InvalidAlgorithmParameterException
+     *     if the {@code KDFParameters} is an invalid value
```

Comment:
Be careful here. Do you really want to throw both `NoSuchAlgorithmException` and `InvalidAlgorithmParameterException` here? Suppose there is an impl that is for the algorithm name but does not accept the parameters. Which exception do you want to throw? Is your rule always reliable?

Reply:
In your current implementation, parameters are never checked. IIUC, it will only be used (i.e. passed into the constructor of implementations) in deriveXyz calls.

This brings out another issue. When deriveXyz is called and and InvalidAlgorithmParameterException is thrown, do we need if it's because the constructor fails or the engineDeriveXyz call fails? This is a bigger problem.